Baskerville/llama2-inference-srun.md

```diff
@@ -0,0 +1,344 @@
+# Llama 2 inference on Baskerville using `srun`
+
+Before you can make use of them you’ll need to download the models.
+See [llama-download.md](llama-download.md) for how to go about doing this on Baskerville.
+
+# Llama 2 7B inference, single node, single GPU
+
+The 7B parameter Llama 2 model will run with a single process on a single node with a single GPU.
+The checkpoint provided for use with the example will, in fact, only work with a single process.
+
+We found at least 16 GiB of RAM was needed to run the model.
+With too little RAM the process will be sent a SIGKILL termination signal, the process will end with an exit code of -9 and an OOM report will be generated.
+
+The RAM, node, GPU and process quantities must be explicitly configured using `srun` for things to work.
+
+## 1. Log in to a single node with a single GPU and a suitable configuration:
+
+You’ll need to replace `<PROJECT_ID>` with the name of your project.
+
+```shell
+$ srun --qos turing \
+    --account <PROJECT_ID> \
+    --time 0:30:0 \
+    --nodes 1 \
+    --gpus 1 \
+    --mem 16384 \
+    --pty bash
+```
+
+## 2. Set up the environment on the node:
+
+Note that the very first command here moves in to the `llama` directory, which is the directory containing the `example_text_completion.py` file.
+We have to be inside this directory for things to work.
+
+```shell
+$ cd llama
+$ module purge
+$ module load baskerville
+$ module load bask-apps/live
+$ module load PyTorch/2.1.2-foss-2022b-CUDA-11.8.0
+$ python -m venv venv
+$ . ./venv/bin/activate
+$ pip install pip --upgrade
+$ pip install -e .
+$ pip install -r requirements.txt
+```
+
+## 3. Run the example script that performs a bunch of inference steps:
+
+```shell
+$ python -m torch.distributed.run \
+    --standalone \
+    --nnodes 1 \
+    --nproc_per_node 1 \
+    example_chat_completion.py \
+    --ckpt_dir llama-2-7b-chat/ \
+    --tokenizer_path tokenizer.model \
+    --max_seq_len 512 \
+    --max_batch_size 6
+```
+
+You should see some output that looks something like this:
+
+```text
+[...]
+Loaded in 20.01 seconds
+[...]
+==================================
+
+System: Always answer with Haiku
+
+User: I am going to Paris, what should I see?
+
+> Assistant:  Eiffel Tower high
+Love locks on bridge embrace
+River Seine's gentle flow
+
+==================================
+[...]
+```
+
+One things to note is that in the Llama docs it [recommends](https://github.com/meta-llama/llama?tab=readme-ov-file#quick-start) to use `torchrun` to execute the script.
+When `torchrun` is used directly in this way it doesn’t support Python virtual environments because it defaults to the system-installed Python.
+To circumvent this we’ve switched `torchrun` out for `python -m torch.distributed.run` instead.
```

Comment:
Neat, I didn't know a) that that was an issue or b) that there was a workaround.

Reply:
I must have hit it as a problem, but it does seem a little odd!